src/LinearMaps.jl

```diff
@@ -61,6 +66,13 @@ function check_dim_mul(Y::AbstractMatrix, A::LinearMap, X::AbstractMatrix)
     return nothing
 end
 
+# check dimension consistency for left multiplication x = y'*A
```

Comment:
What if we relax the "output" type of `mul!` to `LinearAlgebra.AbstractVecOrMat`, and make `check_dim_mul` check all the usual `size` conditions as in line 62 above? We would need only one
```julia
function check_dim_mul(Y::AbstractVecOrMat, A::LinearMap, X::AbstractVecOrMat)
    # @info "checked matrix dimensions" # uncomment for testing
    m, n = size(A)
    (m == size(Y, 1) && n == size(X, 1) && size(Y, 2) == size(X, 2)) || throw(DimensionMismatch("mul!"))
     return nothing
 end
```
First of all, I don't think these checks are expensive. And second, for `X,Y::AbstractVector`s, and `size(Y, 2) == size(X, 2))` may get constant-propagated and compiled away, since they return `1` in that case. Below, we could have `y::AbstractVecOrMat` in all the `mul!`s (which would make your `VecOut` obsolete), and we would need the `VecIn` only in a few cases.

This is not tested and probably not thought through well, but what came to my mind immediately. I would hope that we wouldn't have any issues continuing supporting Julia v1.0.

Reply:
Thanks for the reply and suggestions.  I will take a stab at that and get back to you when done or stuck.